bgp: adding internal reconciliation metrics

harsimran-pabla opened the pull request:
Introducing internal reconciliation metrics which can be used to raise alarms when BGP subsystem is reporting errors or taking longer than usual time to reconcile.

New metrics
- reconcile_error_count ( counter )
- reconcile_run_duration_seconds ( histogram )

reconcile_run_duration_seconds_count metric can be used to track number of reconciliation events in BGP sub-system.

<img width="3189" alt="bgp_reconcile_prom" src="https://github.com/user-attachments/assets/24763781-c9f6-4f47-b8d0-31c027cb26de">


```release-note
BGP: Introducing metrics for tracking health of BGP subsystem reconcile loop
```

harsimran-pabla committed:
bgp: adding internal reconciliation metrics

Introducing internal reconciliation metrics which can be used to raise
alarms when BGP subsystem is reporting errors or taking longer than
usual time to reconcile.

New metrics
- reconcile_error_count ( counter )
- reconcile_run_duration_seconds ( histogram )

reconcile_run_duration_seconds_count metric can be used to track number
of reconciliation events in BGP sub-system.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

rastislavs commented on pkg/bgpv1/manager/metrics.go:
```diff
@@ -0,0 +1,37 @@
+	// ReconcileErrorCount is the number of errors during reconciliation of the BGP manager.
+	ReconcileErrorCount metric.Vec[metric.Counter]
+
+	// ReconcileRunDuration is measures the duration of the reconciliation run. Histogram can
```
```suggestion
	// ReconcileRunDuration measures the duration of the reconciliation run. Histogram can
```

rastislavs reviewed (approved):
Thanks!

harsimran-pabla committed:
bgp operator: add internal error metric

Introducing BGP operator internal reconciliation error metric which can
be used to raise alarm if BGP control plane operator is unhealthy.

Metric
- cilium_operator_bgp_control_plane_cluster_config_error_count

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

qmonnet reviewed (approved):
Looks good to me, thank you!

derailed commented on operator/pkg/bgpv2/metrics.go:
```diff
@@ -0,0 +1,27 @@
+	"github.com/cilium/cilium/pkg/metrics/metric"
+)
+
+type BGPOperatorMetrics struct {
```
nti: docs

derailed commented on operator/pkg/bgpv2/metrics.go:
```diff
@@ -0,0 +1,27 @@
+	BGPClusterConfigErrorCount metric.Vec[metric.Counter]
+}
+
+func NewBGPOperatorMetrics() *BGPOperatorMetrics {
```
nit: docs

derailed reviewed (approved):
@harsimran-pabla Nice work!
